base/abstractdict.jl

```diff
@@ -502,108 +502,149 @@ push!(t::AbstractDict, p::Pair) = setindex!(t, p.second, p.first)
 push!(t::AbstractDict, p::Pair, q::Pair) = push!(push!(t, p), q)
 push!(t::AbstractDict, p::Pair, q::Pair, r::Pair...) = push!(push!(push!(t, p), q), r...)
 
-# hashing objects by identity
+
+# Hashing objects by identity
+_tablesz(x::Integer) = x < 16 ? 16 : one(x)<<((sizeof(x)<<3)-leading_zeros(x-1))
 
 """
-    ObjectIdDict([itr])
+    IdDict([itr])
 
-`ObjectIdDict()` constructs a hash table where the keys are (always)
-object identities.  Unlike `Dict` it is not parameterized on its key
-and value type and thus its `eltype` is always `Pair{Any,Any}`.
+`IdDict{K,V}()` constructs a hash table using object-id as hash and
+`===` as equality with keys of type `K` and values of type `V`.
 
 See [`Dict`](@ref) for further help.
 """
-mutable struct ObjectIdDict <: AbstractDict{Any,Any}
+mutable struct IdDict{K,V} <: AbstractDict{K,V}
     ht::Vector{Any}
     ndel::Int
-    ObjectIdDict() = new(Vector{Any}(uninitialized, 32), 0)
+    IdDict{K,V}() where {K, V} = new{K,V}(Vector{Any}(uninitialized, 32), 0)
 
-    function ObjectIdDict(itr)
-        d = ObjectIdDict()
+    function IdDict{K,V}(itr) where {K, V}
+        d = IdDict{K,V}()
         for (k,v) in itr; d[k] = v; end
         d
     end
 
-    function ObjectIdDict(pairs::Pair...)
-        d = ObjectIdDict()
+    function IdDict{K,V}(pairs::Pair...) where {K, V}
+        d = IdDict{K,V}()
+        sizehint!(d, length(pairs))
+        for (k,v) in pairs; d[k] = v; end
+        d
+    end
+
+    IdDict{K,V}(d::IdDict{K,V}) where {K, V} = new{K,V}(copy(d.ht))
+    function IdDict{K,V}(d::IdDict{K}) where {K, V}
+        d = IdDict{K,V}()
+        sizehint!(d, length(pairs))
         for (k,v) in pairs; d[k] = v; end
         d
     end
+end
+const _IdDict = IdDict{Any,Any} # this is needed to make src/dump.c and src/staticdata.c work
+
+IdDict() = IdDict{Any,Any}()
+IdDict(kv::Tuple{}) = IdDict()
+
+IdDict(ps::Pair{K,V}...)           where {K,V} = IdDict{K,V}(ps)
```

Comment:
We probably don't need all this; we can just default to `Any,Any`.

Reply:
That will make IdDict construction different from Dict construction. Is that desired?

Reply:
It already is; previously all `ObjectIdDicts` were `AbstractDict{Any,Any}`s. ObjectIdDicts are also often heterogeneous, and I don't see a need to break those use cases (i.e. where different-typed keys are added to a dict after construction).

Reply:
I think it would make Julia more consistent (c.f. #20402) to have IdDict constructors behave like Dict constructors (and also like Arrays, and presumably many others which try to guess a tight type parameter according to the arguments).  Also note that of the uses in Base, only one would need updating to`IdDict{Any,Any}(...)`.

That would also make a possible future transition to a more unified dict-backend along the lines of #24932 easier.